```solidity
contract Library {
    function initWallet() {
        // code
    }

    function addOwner(address owner) {
        // code
    }

    function isOwner(address _addr) constant returns (bool) {
        // code
    }
}

contract Wallet {
   function isOwner(address _addr) constant returns (bool) {
       return _walletLibrary.delegatecall(msg.data);
   }

    function() {
        _walletLibrary.delegatecall(msg.data)
    }

}
```

## 띄어놓고 보면 문제가 없는데, 합치면 문제가 생긴다

1. Fallback 함수 → 현재 컨트랙트에 존재하지 않는 메서드를 호출할 경우 실행되는 로직
2. Delegate Call → 현재 컨트랙트의 문맥(`msg` 등)을 그대로 가지고서, 다른 컨트랙트에서 동일한 이름의 함수를 호출

위의 두 가지 기능이 맞물리면서, 의도하지 않은 일들이 발생한다:

- 해당 지갑 컨트랙트에서 사용하지 않지만, 해당 지갑 내의 Fallback 함수 내에서 Delegate Call이 이루어지는 라이브러리 컨트랙트에 존재하는 메서드에 접근 (`.delegatecall()`)
- 외부 라이브러리의 메서드 중, 지갑 컨트랙트의 소유권 등을 조작할 수 있는 메서드의 접근
  - 위의 경우, `.addOwner()`와 같은 메서드 (해당 컨트랙트의 소유권 수정)
- 성공적으로 이러한 취약점이 적용된다면, 토큰이 모두 털리는 등의 사고로 이어질 수 있다
- 문제는 이것이 문법적으로 아무 오류도 없다는 것

## DAO보다 더 광범위한 취약점

- DAO는 하나의 DAO 컨트랙트에 대하여만 문제가 되는 것이므로, *단일한 ETH 자원* 에만 문제가 발생하는 것이다.
  - 해당 DAO 이외에는 안전
- 하지만 Parity 공격의 경우 *외부 라이브러리(컨트랙트)를 사용하는 지갑 컨트랙트* 는 모두 취약점에 노출될 수 있으므로, 수많은 지갑들이 모두 위험에 노출되는 것이다

### Parity Attack으로부터 구제해준다는 사람들

이러한 취약점이 노출된 후, 취약점에 악용되어 ETH가 털리기 전에, ETH를 보관해주겠다고 하는 사람들이 있었다. 돌려받았는지는 미지수...
